File: components/offline_items_collection/core/android/offline_item_bridge_unittest.cc
// Copyright 2020 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "components/offline_items_collection/core/android/offline_item_bridge.h"

#include "base/android/jni_android.h"
#include "testing/gtest/include/gtest/gtest.h"

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "components/offline_items_collection/core/native_j_unittests_jni_headers/OfflineItemBridgeUnitTest_jni.h"

using base::android::AttachCurrentThread;

namespace offline_items_collection {
namespace android {
namespace {

// Tests the JNI bridge that creates Java OfflineItem.
class OfflineItemBridgeTest : public ::testing::Test {
 public:
  OfflineItemBridgeTest()
      : j_test_(JOfflineItemBridgeUnitTestClass::Constructor(
            AttachCurrentThread())) {}

  const jni_zero::ScopedJavaGlobalRef<JOfflineItemBridgeUnitTest>& j_test() {
    return j_test_;
  }

 private:
  jni_zero::ScopedJavaGlobalRef<JOfflineItemBridgeUnitTest> j_test_;
};

// Verfies a default offline item can be created in Java.
TEST_F(OfflineItemBridgeTest, CreateOfflineItem) {
  OfflineItem item;
  auto* env = AttachCurrentThread();
  auto j_offline_item = OfflineItemBridge::CreateOfflineItem(env, item);
  j_test()->testCreateDefaultOfflineItem(env, j_offline_item);
}

}  // namespace
}  // namespace android
}  // namespace offline_items_collection

DEFINE_JNI(OfflineItemBridgeUnitTest)

File: components/offline_items_collection/core/android/native_java_unittests/src/org/chromium/components/offline_items_collection/bridges/OfflineItemBridgeUnitTest.java
// Copyright 2020 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.components.offline_items_collection.bridges;

import org.jni_zero.CalledByNative;
import org.junit.Assert;

import org.chromium.components.offline_items_collection.FailState;
import org.chromium.components.offline_items_collection.OfflineItem;
import org.chromium.components.offline_items_collection.OfflineItemFilter;
import org.chromium.components.offline_items_collection.OfflineItemState;
import org.chromium.components.offline_items_collection.PendingState;

/**
 * Unit test to verify {@link OfflineItemBridge} can create {@link OfflineItem} correctly through
 * JNI bridge from native. See native unit test:
 * (components/offline_items_collection/core/android/offline_item_bridge_unittest.cc).
 */
public class OfflineItemBridgeUnitTest {
    @CalledByNative
    private OfflineItemBridgeUnitTest() {}

    @CalledByNative
    public void testCreateDefaultOfflineItem(OfflineItem item) {
        // Verifies key fields for a default offline item.
        Assert.assertNotNull(item);
        Assert.assertEquals(OfflineItemFilter.OTHER, item.filter);
        Assert.assertFalse(item.isTransient);
        Assert.assertEquals(OfflineItemState.COMPLETE, item.state);
        Assert.assertEquals(FailState.NO_FAILURE, item.failState);
        Assert.assertEquals(PendingState.NOT_PENDING, item.pendingState);
    }
}
